### PR TITLE
Use bf16 as default only if supported

### LINKED ISSUE #327: Error: cutlassF: no kernel found to launch
Encounter an error while trying to run Vicuna model:

`
:~/workspace/lit-gpt$ python chat/base.py --checkpoint_dir checkpoints/lmsys/vicuna-7b-v1.3
Loading model 'checkpoints/lmsys/vicuna-7b-v1.3/lit_model.pth' with {'org': 'lmsys', 'name': 'vicuna-7b-v1.3', 'block_size': 2048, 'vocab_size': 32000, 'padding_multiple': 64, 'padded_vocab_size': 32000, 'n_layer': 32, 'n_head': 32, 'n_embd': 4096, 'rotary_percentage': 1.0, 'parallel_residual': False, 'bias': False, 'n_query_groups': 32, 'shared_attention_norm': False, '_norm_class': 'RMSNorm', 'norm_eps': 1e-06, '_mlp_class': 'LLaMAMLP', 'intermediate_size': 11008, 'condense_ratio': 1}
>> Prompt: how is the weather?
>> Reply: Traceback (most recent call last):
  File "/home/jovyan/workspace/lit-gpt/chat/base.py", line 302, in <module>
    CLI(main)
  File "/opt/saturncloud/envs/saturn/lib/python3.9/site-packages/jsonargparse/_cli.py", line 85, in CLI
    return _run_component(component, cfg_init)
  File "/opt/saturncloud/envs/saturn/lib/python3.9/site-packages/jsonargparse/_cli.py", line 147, in _run_component
    return component(**cfg)
  File "/home/jovyan/workspace/lit-gpt/chat/base.py", line 190, in main
    tokens_generated = decode(fabric, tokenizer, y)
  File "/home/jovyan/workspace/lit-gpt/chat/base.py", line 109, in decode
    for token in token_stream:
  File "/opt/saturncloud/envs/saturn/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 35, in generator_context
    response = gen.send(None)
  File "/home/jovyan/workspace/lit-gpt/chat/base.py", line 61, in generate
    logits = model(idx.view(1, -1), max_seq_length, input_pos)
  File "/opt/saturncloud/envs/saturn/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/saturncloud/envs/saturn/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/saturncloud/envs/saturn/lib/python3.9/site-packages/lightning/fabric/wrappers.py", line 116, in forward
    output = self._forward_module(*args, **kwargs)
  File "/opt/saturncloud/envs/saturn/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/saturncloud/envs/saturn/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/jovyan/workspace/lit-gpt/lit_gpt/model.py", line 107, in forward
    x, self.kv_caches[i] = block(x, (cos, sin), max_seq_length, mask, input_pos, self.kv_caches[i])
  File "/opt/saturncloud/envs/saturn/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/saturncloud/envs/saturn/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/jovyan/workspace/lit-gpt/lit_gpt/model.py", line 168, in forward
    h, new_kv_cache = self.attn(n_1, rope, max_seq_length, mask, input_pos, kv_cache)
  File "/opt/saturncloud/envs/saturn/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/saturncloud/envs/saturn/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/jovyan/workspace/lit-gpt/lit_gpt/model.py", line 247, in forward
    y = self.scaled_dot_product_attention(q, k, v, mask=mask)
  File "/home/jovyan/workspace/lit-gpt/lit_gpt/model.py", line 273, in scaled_dot_product_attention
    return torch.nn.functional.scaled_dot_product_attention(
RuntimeError: cutlassF: no kernel found to launch!`

Maybe is related to some installation issue?
`ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
s3fs 2022.3.0 requires fsspec==2022.3.0, but you have fsspec 2023.6.0 which is incompatible.`

### PR DESCRIPTION
**Pull request by Andrei-Aksionov**
Hi there 👋 

This PR addresses the issue when `bf16-true` is used by default even for cards that don't support it: Nvidia T4 and V100 for example (which are quite popular).

Why for the training `bf16-mixed` is used and not `bf16-true` you may ask?
As @carmocca stated:
> there's many results across tutorials and blogposts that use the current default, so changing this will invalidate them if they weren't explicitly specifying the precision

---

Tested on T4 and A10G: `16-true/mixed` and `bf16-true/mixed` respectively are applied by default.
CPU also by default uses `bf16` precision.

Fixes #327

### CODE REVIEW

File: chat/base.py [hunk 1]
```diff
@@ -139,6 +139,12 @@ def main(
             for more details, see https://github.com/Lightning-AI/lit-gpt/blob/main/tutorials/quantize.md
         precision: Indicates the Fabric precision setting to use.
     """
+    if not precision:
+        if not torch.cuda.is_available() or torch.cuda.is_bf16_supported():
+            precision = "bf16-true"
+        else:
+            precision = "16-true"
```
Comment (carmocca) [inline chat/base.py:146]: It'd be worth adding an utility function for this

File: generate/adapter.py [hunk 2]
```diff
@@ -31,7 +31,7 @@ def main(
     temperature: float = 0.8,
     strategy: str = "auto",
     devices: int = 1,
-    precision: str = "bf16-true",
+    precision: str = None,
```
Comment (carmocca) [inline generate/adapter.py:34]: Optional now

File: generate/base.py [hunk 3]
```diff
@@ -99,7 +99,7 @@ def main(
     quantize: Optional[Literal["bnb.nf4", "bnb.nf4-dq", "bnb.fp4", "bnb.fp4-dq", "bnb.int8", "gptq.int4"]] = None,
     strategy: str = "auto",
     devices: int = 1,
-    precision: str = "bf16-true",
+    precision: str = None,
```
Comment (carmocca) [inline generate/base.py:102]: Optional now

**Review by carmocca — COMMENTED** [review]
Tagging https://github.com/Lightning-AI/lightning/issues/12895 for future reference, as this would be an example

File: generate/adapter.py (hunk 2 above)
Comment (Andrei-Aksionov) [reply generate/adapter.py:34]: How did I miss that.
Thanks

**Comment by rasbt** [comment]
On the note

> Why for the training bf16-mixed is used and not bf16-true you may ask?

I'd say that bf16-mixed is also more robust wrt to convergence in my experience (not specific to Lit-GPT). Actually, I never had a case where mixed-precision resulted in non-convergence (assuming you are coming from 32-true). It's a slower but safer default imho.

**Comment by Andrei-Aksionov** [comment]
Thanks for the info @rasbt !

P.S. As you might have noticed I tend to ask a lot of questions in PRs and this is not because I like to bother people (👿), but because you guys have much more experience then I do so such notes are very useful for me.

**Comment by Andrei-Aksionov** [comment]
I also decided to see what is gonna happen if I provide `bf16-mixed` for the `T4` and in this case there was a clear error message:
> RuntimeError: Current CUDA Device does not support bfloat16. Please switch dtype to float16.

Why it didn't happen for the `bf16-true`? 🤷

File: chat/base.py (hunk 1 above)
Comment (Andrei-Aksionov) [reply chat/base.py]: Done.

File: lit_gpt/utils.py [hunk 4]
```diff
@@ -477,3 +477,20 @@ def chunked_cross_entropy(
         for logit_chunk, target_chunk in zip(logit_chunks, target_chunks)
     ]
     return torch.cat(loss_chunks).mean()
+
+
+def get_default_supported_precision(training: bool, tpu: bool = False):
```
Comment (carmocca) [inline lit_gpt/utils.py:482]:
```suggestion
def get_default_supported_precision(training: bool, tpu: bool = False) -> str:
```

**Review by carmocca — APPROVED** [review]
LGTM, but I'd like to get @awaelchli thoughts first

**Comment by rasbt** [comment]
> hy didn't it happen for the bf16-true?

It's kind of weird but I can also do `tensor([1.], dtype=torch.bfloat16)` on mac now even though it shouldn't be supported afaik (but I may be wrong). I think it's a recent torch nightly thing 
Screenshot 2023-08-15 at 4 19 50 PM

**Comment by awaelchli** [comment]
@rasbt bfloat16 on cpu and bfloat16 on cuda are two different things :))

**Comment by rasbt** [comment]
Oh interesting @awaelchli . I thought the fact that it works on CPU was a bug :D

**Review by awaelchli — APPROVED** [review]
Accept because probably most scripts will work fine with most precision types, but I remain skeptical. 

Also note that Lightning will "downgrade" 16-mixed on CPU to bf16-mixed with a warning (which means no grad scaler, just 32-bit weights + autocast ops to bfloat16). Just so you are aware of it.

**Comment by awaelchli** [comment]
@rasbt On cpu, it's for testing and debugging, but I doubt anyone can take advantage of that for realistic use cases. The speed-up comes from hardware support, hence the GPU requirements.

**Pull request merged**